### ISSUE TITLE
`ps` command doesn't update if the user killed the process somehow.

### ISSUE THREAD

**Issue by ChrisVilches** [ChrisVilches/horsetab#15]
If the user kills the process, the process manager has no way to update the output of the `ps` command.

**Expected:** If the user kills the process, the process manager should react to that, and update the current processes. It doesn't look easy to implement.

## Idea

At the moment of executing `ps` (or periodically) when can traverse the array of children (or some object that contains child information) and see if it's alive. Then update the data accordingly. This seems doable.

**Comment by ChrisVilches** [comment]
Done, most test.

Garbage collector runs every few seconds (I think I should change the frequency) and sets the status to processes. Also prints a message (stdout) saying it ended abnormally. The process isn't removed from the hash map. Just the status is updated.

So now in the `ps` you can see something like this:

Before killing:

```
PID            TIME (s)       STATUS                   COMMAND
241405         2              Running                  toggle_then_sleep_then_toggle 30
```

After `kill 241405` command (but it may take a few seconds to get reflected).

```
PID            TIME (s)       STATUS                   COMMAND
241405         29             signal: 15 (SIGTERM)     toggle_then_sleep_then_toggle 30
```

Then, the main process will consider the process ended, but this isn't guaranteed.

This happened several seconds later after the kill. It seems that it keeps listening to stdout/stderr, not realizing the process has been killed.

```
Done in 61s (signal: 15 (SIGTERM))
```

Maybe I could remove this message? Since it's confusing and doesn't really tells us anything, other than that it's not stuck listening to stdout/stderr anymore.

Maybe we could check if the process has been garbage collected, and if it has, then don't print that message. The only con is that we can't kill the thread that's listening to stdout/stderr, so we just have to let it stay there as a zombie (and eventually end maybe). The worst case scenario is that the user has a few threads stuck not doing anything.

**Comment by ChrisVilches** [comment]
Try another approach using this crate: https://github.com/oconnor663/os_pipe.rs

**Comment by ChrisVilches** [comment]
Done with a different approach. The new approach is much simpler (it also removes several mutexes).

I noticed that the `child.wait()` bit executes correctly even if the output redirection threads are stuck. So I decided to just use that to handle the process exit, and update the process status (in the internal process manager) using that exit status.

It still waits for the output to finish (when it receives `EOF`, which I don't think it's guaranteed it'll always happen).

This works seamlessly without doing any weird hack. Still haven't tested on Windows, but I guess I'll keep this approach even if it doesn't work properly on Windows, since it doesn't even work 100% properly on Linux either, and I don't think there's a much better way that's also simple to implement.

Test & code review.

- [x] The main file is: `src/server/process_manager.rs`
